pkg/server/api_conversion.go

```diff
@@ -0,0 +1,325 @@
+			return nil, fmt.Errorf("conversion strategy %q is not supported for CRD %s", f.crd.Spec.Conversion.Strategy, f.crd.Name)
+		default:
+			return nil, fmt.Errorf("unknown conversion strategy %q for CRD %s", f.crd.Spec.Conversion.Strategy, f.crd.Name)
+		}
+	}
+
+	converter, err = f.newCELConverter(f.crd, apiConversion)
+	if err != nil {
+		return nil, fmt.Errorf("error creating CEL converter for CRD %s|%s: %w", clusterName, f.crd.Name, err)
+	}
+	f.celConverters[string(f.crd.UID)] = converter
+
+	return converter, nil
+}
+
+func (f *deferredConverter) newCELConverter(
+	crd *apiextensionsv1.CustomResourceDefinition,
+	apiConversion *apisv1alpha1.APIConversion,
+) (*celConverter, error) {
+	structuralSchemas := make(map[string]*structuralschema.Structural)
+
+	// Gather all the structural schemas for easy map-based indexing
+	for _, v := range crd.Spec.Versions {
+		if v.Schema == nil {
+			continue
+		}
+
+		internalJSONSchemaProps := &apiextensionsinternal.JSONSchemaProps{}
+		if err := apiextensionsv1.Convert_v1_JSONSchemaProps_To_apiextensions_JSONSchemaProps(v.Schema.OpenAPIV3Schema, internalJSONSchemaProps, nil); err != nil {
+			return nil, fmt.Errorf("failed converting version %s validation to internal version: %w", v.Name, err)
+		}
+
+		structuralSchema, err := structuralschema.NewStructural(internalJSONSchemaProps)
+		if err != nil {
+			return nil, fmt.Errorf("error getting structural schema for version %s: %w", v.Name, err)
+		}
+
+		// Deep copy because PruneDefaults is mutating
+		structuralSchema = structuralSchema.DeepCopy()
+
+		if err := structuraldefaulting.PruneDefaults(structuralSchema); err != nil {
+			return nil, fmt.Errorf("error pruning defaults for version %s: %w", v.Name, err)
+		}
+
+		structuralSchemas[v.Name] = structuralSchema
+	}
+
+	compiledRules, err := kcpconversion.CompileConversions(apiConversion, structuralSchemas)
+	if err != nil {
+		return nil, fmt.Errorf("error compiling conversion rules: %w", err)
+	}
+
+	return &celConverter{
+		crd:               crd,
+		compiledRules:     compiledRules,
+		structuralSchemas: structuralSchemas,
+		apiConversion:     apiConversion,
+	}, nil
+}
+
+type celConverter struct {
+	crd               *apiextensionsv1.CustomResourceDefinition
+	compiledRules     map[string][]*kcpconversion.CompiledRule
+	structuralSchemas map[string]*structuralschema.Structural
+	apiConversion     *apisv1alpha1.APIConversion
+}
+
+func (c *celConverter) Convert(list *unstructured.UnstructuredList, targetGV schema.GroupVersion) (*unstructured.UnstructuredList, error) {
+	convertedList := &unstructured.UnstructuredList{}
+	for i := range list.Items {
+		original := &list.Items[i]
+		converted := original.DeepCopy()
+		converted.SetAPIVersion(targetGV.String())
+
+		originalVersion := original.GetObjectKind().GroupVersionKind().Version
+		compiledRules := c.compiledRules[originalVersion]
+		for _, rule := range compiledRules {
+			if rule.Err != nil {
+				// TODO(ncdc): what to do here?
```

Comment:
I would expect a rule that couldn't compile to be rejected at admission, and that we couldn't get to this function if there were any entries with `Err` populated.

That would require logic to set status on the APIConversion and handle updates, though.

Reply:
The APIConversion is rejected at admission if any conversion specifications are invalid. i.e. it's not even created. I guess we should never hit this condition?